File: src/main/java/hiiragi283/material/api/capability/material/MaterialHandler.java
package hiiragi283.material.api.capability.material;

import hiiragi283.material.api.material.MaterialStack;
import hiiragi283.material.util.HiiragiUtil;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.nbt.NBTTagList;
import net.minecraftforge.common.util.Constants;
import net.minecraftforge.common.util.INBTSerializable;
import org.jetbrains.annotations.NotNull;

import javax.annotation.ParametersAreNonnullByDefault;

@ParametersAreNonnullByDefault
public class MaterialHandler implements IMaterialHandler, INBTSerializable<NBTTagCompound> {

    protected final MaterialStack @NotNull [] materialStacks;
    protected final int capacity;
    protected boolean canFill = true;
    protected boolean canDrain = true;

    public MaterialHandler(int capacity, int slots) {
        this.capacity = capacity;
        this.materialStacks = new MaterialStack[slots];
    }

    public MaterialHandler(int capacity, MaterialStack ... materialStacks) {
        this.capacity = capacity;
        this.materialStacks = materialStacks;
    }

    @Override
    public MaterialStack @NotNull [] getContents() {
        return materialStacks;
    }

    @Override
    public int getCapacity() {
        return capacity;
    }

    @Override
    public boolean canFill() {
        return canFill;
    }

    @Override
    public boolean canDrain() {
        return canDrain;
    }

    @Override
    public boolean canFillMaterialType(MaterialStack materialStack) {
        return canFill();
    }

    @Override
    public boolean canDrainMaterialType(MaterialStack materialStack) {
        return canDrain();
    }

    @Override
    public int fill(MaterialStack resource, boolean doFill) {
        return canFillMaterialType(resource) ? fillInternal(resource, doFill) : 0;
    }

    public int fillInternal(MaterialStack resource, boolean doFill) {
        //resourceがEMPTYの場合 -> 0
        if (resource.isEmpty()) return 0;
        //doFill = falseの場合 -> 数値だけ返す
        if (!doFill) {
            //各MaterialStackに対して搬入可能かチェック
            for (MaterialStack materialStack : getContents()) {
                //MaterialStackがEMPTY -> 最大値を超えない範囲で搬入
                if (materialStack.isEmpty()) {
                    return Math.min(getCapacity(), resource.amount);
                }
                //materialが同じ場合 -> 最大値を超えない範囲で搬入
                else if (materialStack.equalsMaterial(resource)) {
                    return Math.min(getCapacity() - materialStack.amount, resource.amount);
                }
            }
            //搬入不可能 -> 0
        }
        //doFill = trueの場合 -> 実際に搬入を行う
        else {
            //各MaterialStackに対して搬入可能かチェック
            for (int i = 0; i < getContents().length; i++) {
                MaterialStack materialStack = getContents()[i];
                //MaterialStackがEMPTY -> 最大値を超えない範囲で搬入
                if (materialStack.isEmpty()) {
                    getContents()[i].amount = Math.min(getCapacity(), resource.amount);
                    return getContents()[i].amount;
                }
                //materialが同じ場合 ->
                else if (materialStack.equalsMaterial(resource)) {
                    //toFillは搬入可能な量を表す
                    int toFill = getCapacity() - materialStack.amount;
                    //搬入量 < toFill -> resource.amountの分だけ増やす
                    if (resource.amount < toFill) {
                        getContents()[i].amount += resource.amount;
                        toFill = resource.amount;
                    }
                    //搬入量 >= toFill -> 最大値にする
                    else {
                        getContents()[i].amount = getCapacity();
                    }
                    return toFill;
                }
            }
            //materialがすべて一致しなかった場合 -> 0
        }
        return 0;
    }

    @Override
    public int drain(MaterialStack resource, boolean doDrain) {
        return canDrainMaterialType(resource) ? drainInternal(resource, doDrain) : 0;
    }

    public int drainInternal(MaterialStack resource, boolean doDrain) {
        //resourceがEMPTYの場合 -> 0
        if (resource.isEmpty()) return 0;
        //doDrain = falseの場合 -> 数値だけ返す
        if (!doDrain) {
            //各MaterialStackに対して搬出可能かチェック
            for (MaterialStack materialStack : getContents()) {
                //MaterialStackがEMPTYでない && materialが同じ -> 0を下回らない範囲で搬出
                if (!materialStack.isEmpty() && materialStack.equalsMaterial(resource)) {
                    //搬出したい量
                    int toDrain = Math.min(materialStack.amount, resource.amount);
                    return Math.max(0, toDrain);
                }
            }
            //MaterialStackがEMPTY || materialが同じでない -> 0
        }
        //doDrain = trueの場合 -> 実際に搬出を行う
        else {
            //各MaterialStackに対して搬出可能かチェック
            for (int i = 0; i < getContents().length; i++) {
                MaterialStack materialStack = getContents()[i];
                //MaterialStackがEMPTYでない && materialが同じ -> 0を下回らない範囲で搬出
                if (!materialStack.isEmpty() && materialStack.equalsMaterial(resource)) {
                    //搬出したい量
                    int toDrain = Math.min(materialStack.amount, resource.amount);
                    //残る量
                    int remaining = materialStack.amount - toDrain;
                    //remaining <= 0 -> EMPTYを代入する
                    if (remaining <= 0) {
                        getContents()[i] = MaterialStack.EMPTY;
                    }
                    //remaining >= 0 -> 減らした量を代入する
                    else {
                        getContents()[i].amount = remaining;
                    }
                    return Math.max(0, toDrain);
                }
            }
            //materialがすべて一致しなかった場合 -> 0
        }
        return 0;
    }

    //    INBTSerializable    //

    @Override
    public NBTTagCompound serializeNBT() {
        var tagList = new NBTTagList();
        for (MaterialStack materialStack : getContents()) {
            tagList.appendTag(materialStack.serializeNBT());
        }
        var tag = new NBTTagCompound();
        tag.setTag(HiiragiUtil.MATERIALS, tagList);
        return tag;
    }

    @Override
    public void deserializeNBT(NBTTagCompound nbt) {
        NBTTagList tagList = nbt.getTagList(HiiragiUtil.MATERIALS, Constants.NBT.TAG_COMPOUND);
        for (int i = 0; i < tagList.tagCount(); i++) {
            NBTTagCompound tag = tagList.getCompoundTagAt(i);
            var stack = new MaterialStack(tag);
            getContents()[i] = stack;
        }
    }

}

File: src/main/kotlin/hiiragi283/api/capability/material/MaterialHandler.kt
package hiiragi283.api.capability.material

import hiiragi283.api.capability.CapabilityIO
import hiiragi283.api.capability.IOType
import hiiragi283.api.event.MaterialStackEvent
import hiiragi283.api.material.MaterialStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.tileentity.TileEntity
import net.minecraftforge.common.util.INBTSerializable

class MaterialHandler(
    @JvmField var stack: MaterialStack = MaterialStack.EMPTY,
    @JvmField var capacity: Int
) : IMaterialHandler, INBTSerializable<NBTTagCompound>, CapabilityIO<MaterialHandler> {

    //    IMaterialHandler    //

    override fun getMaterialStack(): MaterialStack = stack

    fun setMaterialStack(stack: MaterialStack): MaterialHandler = also { this.stack = stack }

    override fun getMaterialAmount(): Int = stack.amount

    fun setMaterialAmount(amount: Int): MaterialHandler = also { this.stack.amount = amount }

    override fun getCapacity(): Int = capacity

    fun setCapacity(capacity: Int): MaterialHandler = also { this.capacity = capacity }

    override fun insertMaterial(resource: MaterialStack, simulate: Boolean): MaterialStack {
        //handler内がEMPTYの場合，確定で搬入する
        if (this.isEmpty()) {
            if (!simulate) this.stack = resource
            return resource
        }
        //HiiragiMaterialが異なる場合はEMPTYを返す
        if (!resource.equalsMaterial(this.stack)) return MaterialStack.EMPTY
        //現在のamount + 搬入したいamount <= capacityの場合
        else if (this.getMaterialAmount() + resource.amount <= this.capacity) {
            if (!simulate) {
                this.stack.addAmount(resource.amount)
                resource.amount = 0
            }
            tile?.let {
                MaterialStackEvent.fireEvent(MaterialStackEvent.Insert(stack, it.world, it.pos, this))
            }
            return resource
        }
        //capacity - 現在のAmount = 搬入可能なAmount
        else {
            val amountDiff: Int = this.capacity - this.getMaterialAmount()
            if (!simulate) {
                resource.removeAmount(amountDiff)
                this.stack.addAmount(amountDiff)
            }
            tile?.let {
                MaterialStackEvent.fireEvent(MaterialStackEvent.Insert(stack, it.world, it.pos, this))
            }
            return resource.copy(amount = amountDiff)
        }
    }

    override fun extractMaterial(resource: MaterialStack, simulate: Boolean): MaterialStack {
        //HiiragiMaterialが異なる場合はEMPTY
        return if (!resource.equalsMaterial(this.stack)) MaterialStack.EMPTY
        //現在のamount - 搬出したいamount > 0の場合
        else if (this.getMaterialAmount() - resource.amount > 0) {
            if (!simulate) {
                this.stack.removeAmount(resource.amount)
            }
            tile?.let {
                MaterialStackEvent.fireEvent(MaterialStackEvent.Extract(stack, it.world, it.pos, this))
            }
            resource
        } else {
            //stackをEMPTYにする
            val copy = this.stack.copy()
            if (!simulate) {
                this.stack = MaterialStack.EMPTY
            }
            tile?.let {
                MaterialStackEvent.fireEvent(MaterialStackEvent.Extract(stack, it.world, it.pos, this))
            }
            copy
        }
    }

    //    ICapabilityIO    //

    override var ioType: IOType = IOType.INPUT

    override fun getIOType(): IOType = ioType

    override fun setIOType(type: IOType): MaterialHandler = also { ioType = type }

    //    INBTSerializable    //

    override fun serializeNBT(): NBTTagCompound = stack.serializeNBT()

    override fun deserializeNBT(nbt: NBTTagCompound) {
        stack = MaterialStack(nbt)
    }

    //    Custom    //

    var tile: TileEntity? = null

    fun setTile(tile: TileEntity) = also { this.tile = tile }

}